### ISSUE TITLE
systemd-networkd not running when expected

### ISSUE THREAD

**Issue by iandall** [systemd/systemd#25515]
### systemd version the issue has been seen with

252

### Used distribution

Fedora 36

### Linux kernel version used

6.0.8-200.fc36.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

At multi-user.target, systemd-networkd should be running.

### Unexpected behaviour you saw

Weirdly, this is platform dependent. Using exactly the same image (nfs root) a laptop boots OK. A desktop on the same network booting the same image consistently fails.

In both cases, systemd-networkd starts properly in initrd, and then deactivates before switch-root. In the laptop case, networkd starts again, almost immediately, after switchroot. In the desktop case, networkd never starts again without intervention. You can manually start it OK (and it stays running).

If left not running, the network connection is eventually lost when the lease expires. In the case of IPv6, after about 10 minutes (I'm not sure why it happens quicker with IPv6 - maybe something to do with route advertisements).

 Running `systemctl list-dependencies multi-user.target` shows systemd-networkd is definitely a dependency.

### Additional program output to the terminal or log subsystem illustrating the issue

I have captured logs,  but they don't show much. I have a drop-in file for systemd-networkd to set SYSTEMD_LOG_LEVEL=debug but that doesn't change the first run (inside initrd) and the run which doesn't happen after switch-root doesn't happen! In short, the logs look fine, except for the absence of systemd-networkd being started after switch-root.

I don't know how this can happen and satisfy the dependency, but there is no timeout or unusual delay. I can only think that somehow systemd thinks networkd is still active when it is not, or perhaps a timing issue where networkd is still shutting down.

I'm running systemd 250.8-1 from Fedora 36 but in an attempt to solve the problem, changed to systemd-networkd, systemd-network-wait-online and systemd-network-generator built from main/HEAD.

**Comment by yuwata** [comment]
Is `systemd-networkd.service` is enabled on the host?

**Comment by iandall** [comment]
Yes
# systemctl is-enabled systemd-networkd.service
enabled
# systemctl is-enabled systemd-networkd.socket 
enabled

**Comment by iandall** [comment]
I have thought to do a git bisect, however the earliest version I could get to compile still exhibited the same behaviour. I then tried that same version on Fedora 35 (rebuilt on Fedora 35 as well) and it didn't exhibit the behaviour, so I doubt it is systemd itself.

I found that after booting to multi-user, and confirming that systemd-networkd is not running, if I do `systemctl isolate multi-user.target` it causes systemd-networkd to start. I'm wondering if it is something to do with how systemd  is shut down and restarted at switchroot.

**Comment by iandall** [comment]
Booting with a Fedora 35 kernel and initrd, but the Fedora 36 root file system (with the modules for the old kernel) systemd-networkd starts after switch-root properly. The systemd and everying inside the initrd is from Fedora 35.

**Comment by iandall** [comment]
I can confirm that it is the version of systemd in initrd which matters, not the version which runs after switch-root. I've managed to complete a git-bisect, building a new initrd at each stage. In order to get each step to compile, at most stages I had to cherry-pick  b3563e13d4. The result is
```
$ git bisect good
3f60e4488c5d52d01af07523cae6fa2f325014c9 is the first bad commit
commit 3f60e4488c5d52d01af07523cae6fa2f325014c9
Author: Yu Watanabe <watanabe.yu+github@gmail.com>
Date:   Wed Jun 30 01:11:07 2021 +0900

    sd-netlink: split sd_netlink_add_match() into two parts
    
    This also makes netlink_slot_disconnect() correctly unref multicast
    groups.

 src/libsystemd/sd-netlink/netlink-internal.h |  13 +++
 src/libsystemd/sd-netlink/netlink-slot.c     |  24 ++----
 src/libsystemd/sd-netlink/sd-netlink.c       | 121 +++++++++++++++------------
 3 files changed, 87 insertions(+), 71 deletions(-)

```

I haven't yet looked into this commit to try understand what is going on.

**Comment by yuwata** [comment]
Thank you for the bisection.

>  b3563e13d4

I cannot find the commit.

BTW, let me confirm. The networkd is not automatically started? Or, started but failed on boot?

**Comment by iandall** [comment]
Sorry the cherry pick should have been https://github.com/systemd/systemd/commit/e7f46ee3ae1cc66a94b293957721d68dc09d7449. The other was a commit of this cherry-pick on a local branch.

The networkd starts successfully in initrd and terminates at switchroot  (similarly to resolved and hostnamed). It is supposed to start again after switchroot but doesn't. systemctl status says it is "Shutting down..."

A  `journalctl -u systemd-networkd.service --no-hostname -o short-monotonic` for a _successful_  boot looks like:

```
-- Journal begins at Tue 2022-11-29 00:44:01 ACDT, ends at Tue 2022-11-29 14:26:07 ACDT. --
[    1.494079] systemd[1]: Starting Network Configuration...
[    1.516029] systemd-networkd[364]: lo: Link UP
[    1.525414] systemd-networkd[364]: lo: Gained carrier
[    1.525553] systemd-networkd[364]: Enumeration completed
[    1.525611] systemd[1]: Started Network Configuration.
[    1.721437] systemd-networkd[364]: eth0: Interface name change detected, renamed to enp3s0.
[    1.969216] systemd-networkd[364]: enp3s0: Link UP
[    5.076951] systemd-networkd[364]: enp3s0: Gained carrier
[    6.310680] systemd-networkd[364]: enp3s0: Gained IPv6LL
[    8.371509] systemd-networkd[364]: enp3s0: DHCPv4 address 10.0.0.2/24 via 10.0.0.254
[   17.153007] systemd[1]: Starting Network Configuration...
[   17.360396] systemd-networkd[1272]: /usr/lib/systemd/network/80-6rd-tunnel.network:18: Unknown key name 'DHCPPrefixDelegation' in section 'Ne>
[   17.385886] systemd-networkd[1272]: enp3s0: Link UP
[   17.385937] systemd-networkd[1272]: enp3s0: Gained carrier
[   17.386102] systemd-networkd[1272]: lo: Link UP
[   17.386132] systemd-networkd[1272]: lo: Gained carrier
[   17.386315] systemd-networkd[1272]: enp3s0: Gained IPv6LL
[   17.386727] systemd-networkd[1272]: Enumeration completed
[   17.386826] systemd[1]: Started Network Configuration.
[   17.413386] systemd-networkd[1272]: enp3s0: DHCPv4 address 10.0.0.2/24, gateway 10.0.0.254 acquired from 10.0.0.254
[   17.413910] systemd-networkd[1272]: enp0s20u4u5: Link UP

```
Also
```
# journalctl -t systemd --no-hostname -o short-monotonic | grep -E '(networkd|Network Configuration)'
[    1.494079] systemd[1]: Starting Network Configuration...
[    1.525611] systemd[1]: Started Network Configuration.
[   10.638510] systemd[1]: systemd-networkd.service: Deactivated successfully.
[   17.153007] systemd[1]: Starting Network Configuration...
[   17.386826] systemd[1]: Started Network Configuration.

```


For an _unsuccessful_ boot , the entry at 17.153007 and subsequently, are missing.

**Comment by yuwata** [comment]
How about systemd-networkd.socket? Is it started correctly?

**Comment by iandall** [comment]
I can no longer reproduce this problem (even using the "bad" initrds I had saved). I am perplexed as to why this would be - it seemed very repeatable all the while I was doing the git bisect. However, it was always weird that only one host was affected. I can only guess that something on the network was triggering the behaviour.

Anyway, I guess this is good news.

**Comment by yuwata** [comment]
OK, let's close then. Feel free to request to reopen this page if you observe the issue again.

**Comment by iandall** [comment]
For completeness, this seems to actually be an issue with nfs connections over ipv6 breaking when systemd-networkd is stopped. So I had the cause and effect mixed up. The second systemd-networkd was indeed not running, but it was because it was the (ipv6) connection was broken (and hence the nfs root fs) rather than the other way around.

This is a dual stack set up and it seems that ipv4 was being used prior to the "bad' commit. Ironically the "bad' commit was probably fixing something to allow ipv6 to work. Before that, the nfs connection was over ipv4. Also, the problem seemed to go away because I had turned off router advertisements in the router, which resulted in ipv4 being chosen for the nfs route.

I'm going to leave this closed and open a new issue because the title and summary of this is no longer appropriate.

**Comment by malt3** [comment]
I'm hitting this issue with Fedora 37, systemd 251 (251.14-2.fc37). This is very flaky. For roughly 50% of boots i'm seeing the following behavior:
networkd runs in the initrd (dracut), but is activated after switch-root. Also dns doesn't work: 

```
[root@fedora ~]# curl google.com
curl: (6) Could not resolve host: google.com
```

```
[root@fedora ~]# ps aux | grep networkd
root        2460  0.0  0.0   3696  1792 ttyS0    S+   07:13   0:00 grep --color=auto networkd
```

```
journalctl -t systemd --no-hostname -o short-monotonic | grep -E '(networkd|Network Configuration)'
[    2.426503] systemd[1]: Listening on systemd-networkd.socket - Network Service Netlink Socket.
[    2.820612] systemd[1]: Starting systemd-networkd.service - Network Configuration...
[    3.097653] systemd[1]: Started systemd-networkd.service - Network Configuration.
[    3.110984] systemd[1]: Starting systemd-networkd-wait-online.service - Wait for Network to be Configured...
[    6.158113] systemd[1]: Finished systemd-networkd-wait-online.service - Wait for Network to be Configured.
[   68.097446] systemd[1]: systemd-networkd.socket: Deactivated successfully.
[   68.097595] systemd[1]: Closed systemd-networkd.socket - Network Service Netlink Socket.
[   68.098500] systemd[1]: systemd-networkd-wait-online.service: Deactivated successfully.
[   68.098689] systemd[1]: Stopped systemd-networkd-wait-online.service - Wait for Network to be Configured.
[   69.161240] systemd[1]: systemd-networkd.service: Deactivated successfully.
[   69.226799] systemd[1]: Listening on systemd-networkd.socket - Network Service Netlink Socket.
[   69.350733] systemd[1]: systemd-networkd-wait-online.service: Bound to unit systemd-networkd.service, but unit isn't active.
[   69.352129] systemd[1]: Dependency failed for systemd-networkd-wait-online.service - Wait for Network to be Configured.
[   69.358268] systemd[1]: systemd-networkd-wait-online.service: Job systemd-networkd-wait-online.service/start failed with result 'dependency'.
```

```
journalctl -u systemd-networkd.service --no-hostname -o short-monotonic --no-pager
[    2.820612] systemd[1]: Starting systemd-networkd.service - Network Configuration...
[    3.096608] systemd-networkd[350]: lo: Link UP
[    3.096971] systemd-networkd[350]: lo: Gained carrier
[    3.097586] systemd-networkd[350]: Enumeration completed
[    3.097653] systemd[1]: Started systemd-networkd.service - Network Configuration.
[    3.229736] systemd-networkd[350]: eth0: Interface name change detected, renamed to enp1s0.
[    3.246483] systemd-networkd[350]: enp1s0: Configuring with /usr/lib/systemd/network/20-wired.network.
[    3.247790] systemd-networkd[350]: enp1s0: Link UP
[    4.236028] systemd-networkd[350]: enp1s0: Gained carrier
[    4.247574] systemd-networkd[350]: enp1s0: DHCPv4 address 10.42.1.100/16, gateway 10.42.0.1 acquired from 10.42.0.1
[    6.155785] systemd-networkd[350]: enp1s0: Gained IPv6LL
```

```
[root@fedora ~]# systemctl is-active "systemd-networkd.service"
inactive
[root@fedora ~]# systemctl is-active "systemd-networkd.socket"
active
[root@fedora ~]# systemctl is-enabled "systemd-networkd.service"
enabled
[root@fedora ~]# systemctl is-enabled "systemd-networkd.socket"
enabled
```

When waiting for ~15 minutes, the networking fixes itself. Is there any way to determine why `systemd-networkd` is inactive after boot?

**Comment by alexanderhaensch** [comment]
I have the same issue with sys-apps/systemd-253.3-r1 and and dracut-057 on gentoo.

```
Jul 13 15:36:21 gentoo systemd[1]: Listening on systemd-networkd.socket.
Jul 13 15:36:21 gentoo systemd[1]: Starting systemd-network-generator.service...
Jul 13 15:36:21 gentoo systemd[1]: Finished systemd-network-generator.service.
Jul 13 15:36:21 gentoo systemd[1]: Reached target network-pre.target.
Jul 13 15:36:22 gentoo systemd[1]: Starting systemd-networkd.service...
Jul 13 15:36:22 gentoo systemd-networkd[991]: lo: Link UP
Jul 13 15:36:22 gentoo systemd-networkd[991]: lo: Gained carrier
Jul 13 15:36:22 gentoo systemd-networkd[991]: Enumeration completed
Jul 13 15:36:22 gentoo systemd[1]: Started systemd-networkd.service.
Jul 13 15:36:22 gentoo systemd[1]: Reached target network.target.
Jul 13 15:36:22 gentoo systemd-networkd[991]: eth2: Interface name change detected, renamed to enp65s0f0np0.
Jul 13 15:36:23 gentoo systemd-networkd[991]: eth0: Interface name change detected, renamed to enp65s0f1np1.
Jul 13 15:36:30 gentoo systemd-networkd[991]: eth0: Interface name change detected, renamed to enp5s0.
Jul 13 15:36:30 gentoo systemd-networkd[991]: eth0: Interface name change detected, renamed to enp5s0d1.
Jul 13 15:36:33 gentoo systemd[1]: Stopped target network.target.
Jul 13 15:36:33 gentoo systemd[1]: Stopped target network-pre.target.
Jul 13 15:36:33 gentoo systemd[1]: systemd-networkd.socket: Deactivated successfully.
Jul 13 15:36:33 gentoo systemd[1]: Closed systemd-networkd.socket.
Jul 13 15:36:33 gentoo systemd[1]: systemd-network-generator.service: Deactivated successfully.
Jul 13 15:36:33 gentoo systemd[1]: Stopped systemd-network-generator.service.
Jul 13 15:36:33 frenchie systemd[1]: systemd-networkd.service: Deactivated successfully.
Jul 13 15:36:33 frenchie systemd[1]: Listening on systemd-networkd.socket.
Jul 13 15:36:33 frenchie systemd[1]: Starting systemd-network-generator.service...
Jul 13 15:36:33 frenchie systemd[1]: systemd-networkd-wait-online.service: Bound to unit systemd-networkd.service, but unit isn't active.
Jul 13 15:36:33 frenchie systemd[1]: Dependency failed for systemd-networkd-wait-online.service.
Jul 13 15:36:33 frenchie systemd[1]: systemd-networkd-wait-online.service: Job systemd-networkd-wait-online.service/start failed with result 'dependency'.
Jul 13 15:36:33 frenchie systemd[1]: Finished systemd-network-generator.service.
Jul 13 15:36:33 frenchie systemd[1]: Reached target network-pre.target.
Jul 13 15:36:35 frenchie systemd[1]: Reached target network.target.
Jul 13 15:36:35 frenchie systemd[1]: Reached target network-online.target.
Jul 13 15:37:31 frenchie dbus-daemon[4151]: [system] Activating via systemd: service name='org.freedesktop.network1' unit='dbus-org.freedesktop.network1.service' requested by ':1.10' (uid=0 pid=13585 comm="networkctl")
Jul 13 15:37:31 frenchie systemd[1]: Starting systemd-networkd.service...
Jul 13 15:37:31 frenchie systemd-networkd[13586]: lo: Link UP
Jul 13 15:37:31 frenchie systemd-networkd[13586]: lo: Gained carrier
Jul 13 15:37:31 frenchie systemd-networkd[13586]: bridge209: netdev ready
Jul 13 15:37:31 frenchie systemd-networkd[13586]: Enumeration completed
Jul 13 15:37:31 frenchie systemd[1]: Started systemd-networkd.service.
Jul 13 15:37:31 frenchie systemd-networkd[13586]: enp65s0f1np1: Configuring with /etc/systemd/network/20.25G.network.
Jul 13 15:37:31 frenchie systemd-networkd[13586]: enp5s0d1: Configuring with /etc/systemd/network/10.209.bridge.network.
Jul 13 15:37:31 frenchie systemd-networkd[13586]: bridge209: Configuring with /etc/systemd/network/12.209.bridge.network.
Jul 13 15:37:31 frenchie systemd-networkd[13586]: bridge209: Link UP
Jul 13 15:37:31 frenchie systemd-networkd[13586]: enp65s0f1np1: Link UP
Jul 13 15:37:31 frenchie systemd-networkd[13586]: enp5s0d1: Link UP
Jul 13 15:37:31 frenchie systemd-networkd[13586]: enp5s0d1: Gained carrier
Jul 13 15:37:31 frenchie systemd-networkd[13586]: bridge209: Gained carrier
Jul 13 15:37:32 frenchie systemd-networkd[13586]: enp65s0f1np1: Gained carrier
Jul 13 15:37:33 frenchie systemd-networkd[13586]: bridge209: Gained IPv6LL
Jul 13 15:37:34 frenchie systemd-networkd[13586]: enp65s0f1np1: Gained IPv6LL
Jul 13 15:37:37 frenchie systemd-networkd[13586]: vethd3dfaee3: Link UP
Jul 13 15:37:37 frenchie systemd-networkd[13586]: vethd3dfaee3: Gained carrier
Jul 13 15:37:37 frenchie systemd-networkd[13586]: veth772893b2: Link UP
Jul 13 15:37:38 frenchie systemd-networkd[13586]: veth772893b2: Gained carrier
Jul 13 15:37:38 frenchie systemd-networkd[13586]: vethf33a6615: Link UP
Jul 13 15:37:38 frenchie systemd-networkd[13586]: vethf33a6615: Gained carrier
Jul 13 15:37:38 frenchie systemd-networkd[13586]: veth856a66b6: Link UP
Jul 13 15:37:39 frenchie systemd-networkd[13586]: veth856a66b6: Gained carrier
Jul 13 15:38:11 frenchie systemd-networkd[13586]: flannel.1: Link UP
Jul 13 15:38:11 frenchie systemd-networkd[13586]: flannel.1: Gained carrier
Jul 13 15:38:13 frenchie systemd-networkd[13586]: flannel.1: Gained IPv6LL
Jul 13 15:38:22 frenchie systemd-networkd[13586]: cni0: Link UP
Jul 13 15:38:22 frenchie systemd-networkd[13586]: cni0: Gained carrier
Jul 13 15:38:22 frenchie systemd-networkd[13586]: cni0: Lost carrier
Jul 13 15:38:22 frenchie systemd-networkd[13586]: vethe2370fa6: Link UP
Jul 13 15:38:22 frenchie systemd-networkd[13586]: vethe2370fa6: Gained carrier
Jul 13 15:38:22 frenchie systemd-networkd[13586]: cni0: Gained carrier
Jul 13 15:38:24 frenchie systemd-networkd[13586]: cni0: Gained IPv6LL
Jul 13 15:38:24 frenchie systemd-networkd[13586]: vethe2370fa6: Gained IPv6LL


```

there is no network connectivity until networkd is triggered via dbus from networkctl after manual login.


The message:  **systemd-networkd-wait-online.service: Bound to unit systemd-networkd.service, but unit isn't active.** looks like the problem, but i do not understand it

**Comment by iandall** [comment]
I think the ipv6 thing I mentioned earlier is another red herring!

Now I think what happens (used to happen) is this:

-  in intitrd, systemd-networkd is supposed to be stopped before initrd-switch-root.target is reached, but isn't for some reason;
- on switch-root, systemd does some serialization magic to transfer state to the new systemd and also sends a kill signal to everything, before execing the new systemd;
- the new systemd de-serializes the state and thinks systemd-networkd is active;
- at some point systemd-networkd dies (due to the earlier signal), systemd gets the SIGCHILD and says that systemd-networkd has successfully deactivated. At this point systemd thinks systemd-network has been successfully activated in the past and has satisfied any dependencies on it so there would be no requirement to start it again.

The root problem is that systemd-networkd is not deactivated prior to initrd-switch-root. You can see this in the logs, because the "deactivated" message comes after the startup message from the new systemd. It is not clear why this should happen, because all the dependencies look to be configured correctly. The configuration is very similar to other services, such as systemd-resolved, which do stop at the right time.

The inconsistent behavior noted earlier might be down to a timing issue. The good news is that it seems to be fixed in current (c109cff9f998876c3b5abd62501132d0fb0f80a2) though it might just be the timing thing again.

**Comment by yuwata** [comment]
> * in intitrd, systemd-networkd is supposed to be stopped before initrd-switch-root.target is reached, but isn't for some reason;

Should be fixed by 75efd16fb001ba19d12362198ba5c44cb6f40d04.

**Comment by alexanderhaensch** [comment]
Thank you for the pointer to the commit. I created a patch and can verify that it is working on 5 affected servers.

https://bugs.gentoo.org/910404

**Comment by iandall** [comment]
> > * in intitrd, systemd-networkd is supposed to be stopped before initrd-switch-root.target is reached, but isn't for some reason;
> 
> Should be fixed by [75efd16](https://github.com/systemd/systemd/commit/75efd16fb001ba19d12362198ba5c44cb6f40d04).

Except 253.5-1.fc38 exhibits the problem and already has those changes in it (perhaps a Fedora patch).
```
# systemctl show | grep '^Version'
Version=253.5-1.fc38
# systemctl show systemd-networkd.service | grep 'initrd-switch-root\.target'
Conflicts=shutdown.target initrd-switch-root.target
Before=systemd-networkd-wait-online.service shutdown.target initrd-switch-root.target network.target multi-user.target
# journalctl -b0 -t systemd --no-hostname --output=short-monotonic | grep '\(networkd\|initrd-switch-root\|running in system mode\)' > /var/tmp/journal.log
```
[journal.log](https://github.com/systemd/systemd/files/12068540/journal.log)

It is less easy to capture the information in initrd, but stopping with the command line option rc.break, the results are the same.

So some other change is/was needed besides that commit. Regardless, the current git HEAD seems to work.
